### PR TITLE
[clang-tidy] BTC-AG module and test added. Sharable-based objects can…

### PR DESCRIPTION
**Pull request by prciesie**
…not be stack instantiated.

### CODE REVIEW

File: clang-tidy/btc/DerivedFromSharableStackProhibitionCheck.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+/// Internal Btc-ag check. Purpose is to safeguard from creating Sharable based
+/// objects on stack.
+///
+/// For the user-facing documentation see:
+/// http://clang.llvm.org/extra/clang-tidy/checks/btc-derived-from-sharable-stack-prohibition.html
+class DerivedFromSharableStackProhibitionCheck : public ClangTidyCheck {
+ public:
+  DerivedFromSharableStackProhibitionCheck(StringRef Name,
+                                           ClangTidyContext *Context)
+      : ClangTidyCheck(Name, Context) {}
```
Comment (sigiesec) [inline clang-tidy/btc/DerivedFromSharableStackProhibitionCheck.h]: Move this to the .cpp file?

File: clang-tidy/btc/DerivedFromSharableStackProhibitionCheck.h [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+#define LLVM_CLANG_TOOLS_EXTRA_CLANG_TIDY_BTC_DERIVEDFROMSHARABLESTACKPROHIBITOIONCHECKCHECK_H
+
+#include "../ClangTidy.h"
+
+namespace clang {
+namespace tidy {
+namespace btc {
+
+/// Internal Btc-ag check. Purpose is to safeguard from creating Sharable based
+/// objects on stack.
```
Comment (sigiesec) [inline clang-tidy/btc/DerivedFromSharableStackProhibitionCheck.h]: I am not completely sure what the best wording would be, but "on stack" is not accurate. In the example below, if GodObject is created on the heap, its field will also be on the heap. The point is that it is not allocated as a *separate* heap object (using CreateAuto/CreateShared/CreateUnique) resp. it is not managed by a smart pointer *directly*.

File: test/clang-tidy/btc-derived-from-sharable-stack-prohibition.cpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+class Sharable {
+ public:
+  Sharable(int value) : val(value) {}
+  int do_Sharable() { return val; }
+
+ private:
+  int val;
+};
+
+class DerivedFromSharable : public Sharable {
```
Comment (sigiesec) [inline test/clang-tidy/btc-derived-from-sharable-stack-prohibition.cpp]: Maybe add another test case that uses virtual inheritance?

File: clang-tidy/btc/DerivedFromSharableStackProhibitionCheck.h (hunk 1 above)
Comment (prciesie) [reply clang-tidy/btc/DerivedFromSharableStackProhibitionCheck.h]: It's done consistently with goole template for checks. All checks look this way

File: test/clang-tidy/btc-derived-from-sharable-stack-prohibition.cpp (hunk 3 above)
Comment (prciesie) [reply test/clang-tidy/btc-derived-from-sharable-stack-prohibition.cpp]: Ok, will do.

File: clang-tidy/btc/DerivedFromSharableStackProhibitionCheck.h (hunk 1 above)
Comment (sigiesec) [reply clang-tidy/btc/DerivedFromSharableStackProhibitionCheck.h]: Ok. Then leave it like this :)

**Comment by sigiesec** [comment]
@prciesie Did you work on the comments? There was no update in the branch, maybe you just forgot to push?

**Comment by sigiesec** [comment]
@prciesie Can you please fix the open review comments, so this can be merged?

**Review by sigiesec — APPROVED** [review]

**Pull request merged**